Updates for Django 1.9 and improvements to README.md

YPCrumble opened the pull request:
@skorokithakis I was investigating the Django 1.9 error in the issue you just closed and went through and updated a few more areas for Django 1.9. This fixes a bug running the test suite on Django 1.9 via `python test.py`. 

Also added links in the README.md to make it more easily navigable, and added a TOC entry for `get_object_or_this`.

This update to `test.py` now works on Django 1.6+, I didn't write any fallbacks for older versions of Django.

I'm really new to open-source development so I apologize if this isn't the correct protocol. Please let me know if I need to adjust anything or if these changes aren't a good fit?

skorokithakis commented:
Thanks for this, it looks pretty good! I'm not very thrilled about breaking the tests on older versions, do you know if the fallback would be a lot of effort to write? I'd like to keep supporting older versions if possible, and tests are a part of it, so I'd prefer that they worked on those as well.

skorokithakis commented on annoying/tests/urls.py:
```diff
@@ -1,13 +1,13 @@
 from __future__ import absolute_import
 
-from django.conf.urls import patterns
+from django.conf.urls import url
```
This would be pretty easy to make backwards-compatible like so:

```
try:
    from django.conf.urls import patterns
except ImportError:
    # Hack for backwards-compatibility.
    patterns = lambda x: list(x[1:])

urlpatterns = patterns('',
    <etc>
```

skorokithakis commented on test.py:
```diff
@@ -16,27 +16,31 @@
     settings.configure(**SETTINGS)
 
-from django.db import models
-from django.conf.urls import patterns
```
Same as above for this.

YPCrumble commented:
@skorokithakis just updated per your suggestion. I added a note that Django-annoying supports Django 1.6+. This is in line with other packages like Django-Extensions and even the Django project doesn't support lower than 1.7. The tests weren't working previously for <= Django 1.5 so I didn't make a fix for those :).

I also updated the Travis-CI build script for a matrix of applicable Python and Django versions. I'm also unsure it was actually running the tests properly before - I think they were just passing because the script `python setup.py test` isn't actually registered.

Take a look and let me know what you think!

skorokithakis commented:
That looks great, thank you for fixing this up! One minor issue, you're doing version detection rather than doing feature detection (and duplicating code, as well). You can use the code I wrote in the comment above to only fall back if the import fails, and to avoid repeating the URLs as well. Other than that, I'm very happy that you got the tests to work again.

YPCrumble commented:
@skorokithakis added feature detection for the `patterns` import. Note I used `patterns = lambda *x: list(x[1:])` (with the *x) to account for multiple arguments to the lambda function.

I'm not sure how to do feature detection for Django's `SETTINGS` dict. If you know a way I'd be happy to do that as well.

Thanks for reviewing!

skorokithakis commented:
Oh yes, good call about the asterisk. The settings are probably fine the way you have them, the changes are too different to do any other way. I'm fine with merging this if you are, what do you think?

YPCrumble commented:
One unrelated thing I didn't look at is that setup.py still has a reference
to 0.8.5, not sure if that's an issue for installation or anything if
you're not using pip.

I also didn't adjust the `classifiers` setting in setup.py to remove, e.g.,
Python 2.6. Probably should do that. Does "Python 2" remain if we also have
Python 2.7 as a classifier?

On Mon, Jan 25, 2016 at 8:23 PM, Stavros Korokithakis <
notifications@github.com> wrote:

> Oh yes, good call about the asterisk. The settings are probably fine the
> way you have them, the changes are too different to do any other way. I'm
> fine with merging this if you are, what do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/skorokithakis/django-annoying/pull/47#issuecomment-174757510>
> .
>

skorokithakis commented:
Hmm, where's that in `setup.py`? I don't see it. The classifiers should be adjusted, yes, and I think Python 2 can remain, since 2.7 is there, but I'm not entirely sure.

YPCrumble commented:
Toward the bottom of `setup.py`. Just made a few changes - let me know if
you like those?

On Mon, Jan 25, 2016 at 8:31 PM, Stavros Korokithakis <
notifications@github.com> wrote:

> Hmm, where's that in setup.py? I don't see it. The classifiers should be
> adjusted, yes, and I think Python 2 can remain, since 2.7 is there, but I'm
> not entirely sure.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/skorokithakis/django-annoying/pull/47#issuecomment-174759778>
> .
>

skorokithakis commented:
It looks like you have a conflict now, there's no version in the latest setup.py:

https://github.com/skorokithakis/django-annoying/blob/develop/setup.py

Try merging my develop into yours, that should fix it.

YPCrumble commented:
Whoops! Just did that, thanks.

skorokithakis commented:
Merged, thank you very much for your help!
